proof/refine/ARM/Schedule_R.thy

```diff
@@ -4862,136 +4869,438 @@ lemma refillBudgetCheck_corres:
-                                 apply (rule setSchedulerAction_corres[simplified dc_def])
-                                 apply (wp | simp)+
-
-                             (* isHighestPrio *)
-                             apply (clarsimp simp: if_apply_def2)
-                             apply ((wp (once) hoare_drop_imp)+)[1]
-                            apply (simp add: if_apply_def2)
-                             apply ((wp (once) hoare_drop_imp)+)[1]
-                           apply wpsimp+
-                     apply (wpsimp simp: etcb_relation_def)+
-            apply (rule tcbSchedEnqueue_corres)
-           apply wpsimp+
-
-           apply (clarsimp simp: conj_ac cong: conj_cong)
-           apply wp
-           apply (rule_tac Q="\<lambda>_ s. valid_blocked_except t s \<and> scheduler_action s = switch_thread t"
-                    in hoare_post_imp, fastforce)
-           apply (wp add: tcb_sched_action_enqueue_valid_blocked_except
-                          tcbSchedEnqueue_invs'_not_ResumeCurrentThread thread_get_wp
-                     del: gets_wp)+
-       apply (clarsimp simp: conj_ac if_apply_def2 cong: imp_cong conj_cong del: hoare_gets)
-       apply (wp gets_wp)+
-
-   (* abstract final subgoal *)
+    apply (rule corres_split[OF switchSchedContext_corres])
+      apply (rule corres_split[OF getReprogramTimer_corres])
+        apply (clarsimp simp: when_def)
+        apply (rule corres_split[OF setNextInterrupt_corres])
+          apply (rule setReprogramTimer_corres)
+         apply (wpsimp wp: hoare_vcg_ball_lift2| wps)+
+     apply (rule_tac Q="\<lambda>_.invs" in hoare_post_imp, fastforce)
+     apply wpsimp+
+    apply (rule_tac Q="\<lambda>_. invs'" in hoare_post_imp, fastforce)
+    apply (wpsimp wp: switchSchedContext_invs')
+   apply (clarsimp simp: valid_state_def valid_release_q_def)
+  apply fastforce
+  done
+
+lemma tcb_sched_action_valid_state[wp]:
+  "tcb_sched_action action thread \<lbrace>valid_state\<rbrace>"
+  by (wpsimp simp: tcb_sched_action_def set_tcb_queue_def get_tcb_queue_def valid_state_def
+               wp: hoare_drop_imps hoare_vcg_all_lift)
+
+crunches setQueue, addToBitmap
+  for isSchedulable_bool[wp]: "isSchedulable_bool tcbPtr"
+  (simp: bitmap_fun_defs isSchedulable_bool_def isScActive_def)
+
+lemma tcbSchedEnqueue_isSchedulable_bool[wp]:
+  "tcbSchedEnqueue tcbPtr \<lbrace>isSchedulable_bool tcbPtr'\<rbrace>"
+  apply (clarsimp simp: tcbSchedEnqueue_def unless_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+  apply (rule hoare_when_cases, simp)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+  apply (wpsimp wp: threadSet_wp)
+  apply (fastforce simp: obj_at_simps isSchedulable_bool_def pred_map_simps opt_map_def
+                         isScActive_def
+                  split: option.splits)
+  done
+
+lemma cur_sc_tcb_sc_at_cur_sc:
+  "\<lbrakk>valid_objs s; cur_sc_tcb s\<rbrakk> \<Longrightarrow> sc_at (cur_sc s) s"
```

Comment:
This is nice and general, should we move it to Invariants_H?

Reply:
Sure. Moved

Reply:
Well, let's move it to Invariants_AI